### ISSUE TITLE
Still having problem with the worker thread.

### ISSUE THREAD

**Issue by andresantacruz** [GameAnalytics/GA-SDK-CPP#82]
Hi,

I'm still having problem with the worker thread of the lib.

The thread sometimes just don't die properly upon `ExitProcess`.

What is the correct way to get the thread object so I can assure the thread to be dead before the main thread exits its execution?

**Comment by the1schwartz** [comment]
If you still have issues with the worker thread I can try to expose a method which tells if the thread has finished or not.

**Comment by andresantacruz** [comment]
> 
> 
> Do you call GameAnalytics::onQuit() when quiting the application?

Yes, I call `endSession()` and `onQuit()` when exiting the process. An important note here: I'm loading the lib in another DLL; if I call `onQuit` just before calling `ExitProcess` the GA's worker thread keeps alive whatsoever; strangly the GA's thread only shutdown if I call `onQuit` at DllMain (in PROCESS_DETAH) of my DLL (which is what I'm doing but still get the thread to keep alive sometimes). Does the function `onQuit` blocks execution untill the work is done or it works async?

> 
> 
> If you still have issues with the worker thread I can try to expose a method which tells if the thread has finished or not.

IDK if that will help. I can create a spin loop to wait untill the thread is finished but the problem may be cause by other thing. Can you think of any possible execution flow in method `onQuit()` that doesn't kill the thread after few seconds? Does this method have a timeout mechanism? If not, this could work as a palliative solution while the real source of this problem can't be found. So user of the lib just spin loop after `onQuit` until the GA's thread is finished (which will be done either by the expected execution flow or by timeout).

**Comment by the1schwartz** [comment]
When `onQuit` has been called an endThread flag is set to tell the worker thread to stop. Also see the logic here: https://github.com/GameAnalytics/GA-SDK-CPP/blob/master/source/gameanalytics/GAThreading.cpp#L141-L173

`onQuit` also tries to wait until the worker thread is finished: https://github.com/GameAnalytics/GA-SDK-CPP/blob/master/source/gameanalytics/GameAnalytics.cpp#L903-L928

So in theory is should stop with a couple of seconds. I have not tested the dll scenario but maybe I will test it out to see what goes wrong. If you have a simple test project where this issue is reproducible I will be happy to test with that.

**Comment by andresantacruz** [comment]
Hmm, IDK then :(

One thing I missed in the code snippets you showed is the presence of mutexes. IDK how you are managing the shared objects in the code but I expected to see some locks at least in the code that changes `_endThread`. Is that supposed to be a lock-free code or something?

I can try to create a test application for you but it will take some days as I'm very busy nowadays :(

**Comment by the1schwartz** [comment]
_endThread is of type std::atomic<bool> so there should be no need for mutexes as it can only be accesed by one thread at a time. Ok thanks. I will try to make a project myself in the meantime

**Comment by the1schwartz** [comment]
I have tried to test the SDK using a DLL. I have just tested it in a simple windows desktop application made with C++. You can check it out here: https://drive.google.com/file/d/13O00rEboBO6oh4KNXQ9azz4xDKHNLi7y/view?usp=sharing

You probably need to check the path to the GameAnalytics DLL before building the project. When I tested it, it worked when I called the `onQuit` method when getting the close event for the application. Let me know if you still get the error using the provided test project.

**Comment by xdd12345** [comment]
HI，
    I have the same problem with andresantacruz.The thread sometimes just don't die properly upon ExitProcess.

**Comment by the1schwartz** [comment]
Do you also call onQuit?

**Comment by xdd12345** [comment]
Yes

**Comment by xdd12345** [comment]
> Do you also call onQuit?
yes

**Comment by the1schwartz** [comment]
Ok would you be able to send me a test project where this is reproducible as I can't seem to reproduce it myself

**Comment by xdd12345** [comment]
> Ok would you be able to send me a test project where this is reproducible as I can't seem to reproduce it myself

Sure, I try it.would you give me a e-mail address?

**Comment by the1schwartz** [comment]
Would can send it to sdk@gameanalytics.com

**Comment by andresantacruz** [comment]
@the1schwartz @xdd12345  Any news in the subject?

**Comment by xdd12345** [comment]
> @the1schwartz @xdd12345 Any news in the subject?

Sry,I've been a bit busy lately, forget it，The project has not been sent yet. It will be sent in this week.
